### PR TITLE
base64: encode without snprintf: a 29x speedup

### PR DESCRIPTION
**Pull request by bagder**
For speed

# Test setup

https://gist.github.com/bagder/baee5725e09a5977ef5b0471e2b3334f

## Measurements

Best out  three runs on my machine.

Old code: 5m55.291
New code: 12.254 seconds

A 28.99 times speedup.

### CODE REVIEW

**Review by jay — APPROVED** [review]

File: lib/base64.c [hunk 1]
```diff
@@ -196,51 +195,41 @@ static CURLcode base64_encode(const char *table64,
...
+  while(insize >= 3) {
+    *output++ = table64[ in[0] >> 2 ];
+    *output++ = table64[ ((in[0] & 0x03) << 4) | (in[1] >> 4) ];
+    *output++ = table64[ ((in[1] & 0x0F) << 2) | ((in[2] & 0xC0) >> 6) ];
+    *output++ = table64[ in[2] & 0x3F ];
+    insize -= 3;
+    in += 3;
+  }
+  if(insize) {
+    /* this is only one or two bytes now */
```
Comment (jay) [inline lib/base64.c:204]: at this point you could unroll it and get rid of inputparts and i

```suggestion
    /* this is only one or two bytes now */
    unsigned char ibuf[2];
    ibuf[0] = *in++;
    insize--;
    if(insize) {
      ibuf[1] = *in++;
     --insize;
    }
    else
      ibuf[1] = 0;
```

then later
~~~c
    if(!ibuf[1]) {
      if(*padstr) {
        ...
~~~

Comment (bagder) [reply lib/base64.c:204]: yes, I did that in a subsequent commit!

**Review by jay — APPROVED** [review]

**Comment by Bouska** [comment]
You can remove line 41, we don't need the `curl_printf.h` header anymore.

**Comment by MonkeybreadSoftware** [comment]
Great work.

**Pull request closed without merging**